index.d.ts

```diff
@@ -0,0 +1,60 @@
+export interface Action {
+  type: string;
+}
+
+export type Reducer<S> = (state: S, action: Action) => S;
+
+export type Dispatch = (action: any) => any;
```

Comment:
Tricky situation - JS (and other dynamically typed langs) encourages to build structures that accept anything and handle type inside (runtime reflection) while static typing is about splitting it to handle different inputs/outputs in separate functions/methods.

Thunk (and probably many other) middleware is so elastic, that best solution is to use just:

```typescript
interface DefaultDispatch {
    <A>(action: A) => A;
    <A, B>(action: A) => B;
}
```

And use it with thunk:

```typescript
const dispatch: DefaultDispatch;

…
…  dispatch<any>() …
```